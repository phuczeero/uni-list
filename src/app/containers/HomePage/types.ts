import { University } from 'types/University';

/* --- STATE --- */
export interface HomePageState {
  searchTerm: string;
  loading: boolean;
  universities: University[];
  error: Error | null;
}

export type ContainerState = HomePageState;
