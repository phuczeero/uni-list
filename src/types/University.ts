export interface University {
  alpha_two_code: string;
  web_pages: string[];
  country: string;
  'state-province': string;
  name: string;
  domains: string[];
}
