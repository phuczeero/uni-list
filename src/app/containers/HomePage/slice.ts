import { PayloadAction } from '@reduxjs/toolkit';
import { University } from 'types/University';
import { createSlice } from 'utils/@reduxjs/toolkit';
import { ContainerState } from './types';

// The initial state of the HomePage container
export const initialState: ContainerState = {
  searchTerm: '',
  universities: [],
  loading: false,
  error: null,
};

const homePageSlice = createSlice({
  name: 'homePage',
  initialState,
  reducers: {
    loadUniversities(state, action: PayloadAction<string>) {
      state.searchTerm = action.payload;
      state.loading = true;
    },
    loadSuccess(state, action: PayloadAction<University[]>) {
      state.loading = false;
      state.universities = action.payload;
    },
    loadError(state, action: PayloadAction<Error>) {
      state.error = action.payload;
      state.loading = false;
    },
    saveToJSON() {},
  },
});

export const {
  actions: homePageActions,
  reducer,
  name: sliceKey,
} = homePageSlice;
