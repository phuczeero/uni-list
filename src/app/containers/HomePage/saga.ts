import { call, put, select, takeLatest } from 'redux-saga/effects';
import { saveAs } from 'file-saver';
import { University } from 'types/University';
import { request } from 'utils/request';
import { selectSearchTerm, selectUniversities } from './selectors';
import { homePageActions } from './slice';

export function* loadUniversities() {
  const searchTerm = yield select(selectSearchTerm);
  const requestURL = `http://universities.hipolabs.com/search?name=${searchTerm}`;

  try {
    const universities: University[] = yield call(request, requestURL);
    yield put(homePageActions.loadSuccess(universities));
  } catch (err) {
    yield put(homePageActions.loadError(err));
  }
}

export function* saveToJSON() {
  const universities: University[] = yield select(selectUniversities);
  const data: string = JSON.stringify(universities);
  const blob = new Blob([data], { type: 'application/json;charset=utf-8' });
  saveAs(blob, 'users.json');
}

export function* homePageSaga() {
  yield takeLatest(homePageActions.loadUniversities, loadUniversities);
  yield takeLatest(homePageActions.saveToJSON, saveToJSON);
}
