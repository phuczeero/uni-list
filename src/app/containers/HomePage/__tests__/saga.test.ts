import { put, takeLatest } from 'redux-saga/effects';
import * as slice from '../slice';
import FileSaver from 'file-saver';

import { loadUniversities, homePageSaga, saveToJSON } from '../saga';
import { University } from 'types/University';

jest.mock('file-saver', () => ({ saveAs: jest.fn() }));

describe('loadUniversities Saga', () => {
  let searchTerm: any;
  let universities: any;
  let loadUniversitiesIterator: ReturnType<typeof loadUniversities>;

  beforeEach(() => {
    loadUniversitiesIterator = loadUniversities();

    const selectDescriptor = loadUniversitiesIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should dispatch the loadSuccess action if it requests the data successfully', () => {
    searchTerm = 'test';
    universities = [
      {
        name: 'test',
      },
    ];

    const requestDescriptor = loadUniversitiesIterator.next(searchTerm).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = loadUniversitiesIterator.next(universities).value;
    expect(putDescriptor).toEqual(
      put(slice.homePageActions.loadSuccess(universities)),
    );
  });

  it('should dispatch the response error', () => {
    searchTerm = 'test';
    const error = new Error('some error');

    const requestDescriptor = loadUniversitiesIterator.next(searchTerm).value;
    expect(requestDescriptor).toMatchSnapshot();

    const putDescriptor = loadUniversitiesIterator.throw(error).value;
    expect(putDescriptor).toEqual(put(slice.homePageActions.loadError(error)));
  });
});

describe('saveToJSON Saga', () => {
  const saveToJSONIterator = saveToJSON();

  it('should select universities', () => {
    const selectDescriptor = saveToJSONIterator.next().value;
    expect(selectDescriptor).toMatchSnapshot();
  });

  it('should handle saveAs', () => {
    const universities = [
      {
        name: 'test',
      },
    ] as University[];

    saveToJSONIterator.next(universities);
    expect(FileSaver.saveAs).toBeCalled();
  });
});

describe('homePageSaga Saga', () => {
  const homePageIterator = homePageSaga();
  it('should start task to watch for loadUniversities action', () => {
    const takeLatestDescriptor = homePageIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.homePageActions.loadUniversities, loadUniversities),
    );
  });

  it('should start task to watch for saveToJSON action', () => {
    const takeLatestDescriptor = homePageIterator.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(slice.homePageActions.saveToJSON, saveToJSON),
    );
  });
});
