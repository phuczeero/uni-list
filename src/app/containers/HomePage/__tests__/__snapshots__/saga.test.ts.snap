// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadUniversities Saga should dispatch the loadSuccess action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`loadUniversities Saga should dispatch the loadSuccess action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://universities.hipolabs.com/search?name=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`loadUniversities Saga should dispatch the response error 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;

exports[`loadUniversities Saga should dispatch the response error 2`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [
      "http://universities.hipolabs.com/search?name=test",
    ],
    "context": null,
    "fn": [Function],
  },
  "type": "CALL",
}
`;

exports[`saveToJSON Saga should select universities 1`] = `
Object {
  "@@redux-saga/IO": true,
  "combinator": false,
  "payload": Object {
    "args": Array [],
    "selector": [Function],
  },
  "type": "SELECT",
}
`;
